workers/unity/Packages/io.improbable.gdk.tools/.CodeGenTemplate/CodeGenerationLib/CodeWriter/Scopes/Abstract/ScopeBody.cs

```diff
@@ -0,0 +1,107 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+
+namespace Improbable.Gdk.CodeGeneration.CodeWriter.Scopes
+{
+    /// <summary>
+    /// A ScopeBlock with helpers for writing code found in method bodies.
+    /// </summary>
+    public abstract class ScopeBody : ScopeBlock
+    {
+        protected ScopeBody(string declaration, Action<ScopeBody> populate) : base(declaration)
+        {
+            populate(this);
+        }
+
+        protected ScopeBody(string declaration = null) : base(declaration)
+        {
+        }
+
+        public void WriteLine(string snippet)
+        {
+            Add(new Text(snippet));
+        }
+
+        public void WriteLine(Action<StringBuilder> populate)
+        {
+            var sb = new StringBuilder();
+            populate(sb);
+
+            Add(new Text(sb.ToString()));
+        }
+
+        public void WriteLine(IEnumerable<string> snippets)
+        {
+            if (snippets.Any())
```

Comment:
This _could_ consume the `IEnumerable` (it depends on the implementation). What would happen if you pass in an empty enumerable to `new TextList()`.

Reply:
TextList can handle an empty enumerable - the issue is that adding an empty textlist would have extra inter-codeblock vertical spacing

Reply:
I think that's better than unexpectedly missing output since the user passed in a consumable `IEnumerable<string>`. If we want to keep it this way, it should be noted in the API this quirk/restriction

Reply:
I would rather note the quirk because the ability to call `Line(some_linq)` was invaluable when porting across the generators